kolena/_experimental/dataset/_evaluation.py

```diff
@@ -48,189 +43,173 @@
-        )
-    )
+    df = _fetch_results(dataset, model)
 
+    df_datapoints = _to_deserialized_dataframe(df.drop_duplicates(subset=[COL_DATAPOINT]), column=COL_DATAPOINT)
     eval_configs = df[COL_EVAL_CONFIG].unique()
-    df_by_eval = []
+    df_results_by_eval = []
     for eval_config in eval_configs:
         df_matched = df[df[COL_EVAL_CONFIG] == eval_config if eval_config is not None else df[COL_EVAL_CONFIG].isnull()]
-        df_by_eval.append(
+        df_results_by_eval.append(
             (
                 json.loads(eval_config) if eval_config is not None else None,
-                _to_deserialized_dataframe(df_matched, column=COL_DATAPOINT),
-                _to_deserialized_dataframe(df_matched, column=COL_INFERENCE),
-                _to_deserialized_dataframe(df_matched, column=COL_METRICS),
+                _to_deserialized_dataframe(df_matched, column=COL_RESULT),
             ),
         )
 
-    return df_by_eval
+    return df_datapoints, df_results_by_eval
+
 
+def _validate_configs(configs: List[TYPE_EVALUATION_CONFIG]) -> None:
+    n = len(configs)
+    for i in range(n):
+        for j in range(i + 1, n):
+            if configs[i] == configs[j]:
+                raise IncorrectUsageError("duplicate eval configs are invalid")
 
-def validate_data(left: pd.DataFrame, right: pd.DataFrame) -> None:
+
+def _validate_data(left: pd.DataFrame, right: pd.DataFrame) -> None:
     if len(left) != len(right):
         raise IncorrectUsageError("numbers of rows between two dataframe do not match")
 
 
+def _validate_on(left: pd.DataFrame, right: pd.DataFrame, on: TEST_ON_TYPE) -> None:
+    if on is None:
+        raise IncorrectUsageError("on cannot be None")
+    # works for both string and list of string
+    if len(on) == 0:
+        raise IncorrectUsageError("on cannot be empty")
+
+    if isinstance(on, str):
+        if on not in left.columns or on not in right.columns:
+            raise IncorrectUsageError(f"column {on} doesn't exist in target dataframe")
+    elif isinstance(on, list):
+        for col in on:
+            if col not in left.columns or col not in right.columns:
+                raise IncorrectUsageError(f"column {col} doesn't exist in target dataframe")
+
+
+def _get_single_df_result(
+    df_datapoints: pd.DataFrame,
+    df_result_input: pd.DataFrame,
+    on: TEST_ON_TYPE,
+) -> pd.DataFrame:
+    if not on:
+        return df_result_input
+
+    if isinstance(on, str):
+        on = [on]
+
+    _validate_on(df_datapoints, df_result_input, on)
+    df_result = df_datapoints[on].merge(df_result_input, how="left", on=on)
+    return df_result
+
+
 def test(
     dataset: str,
     model: str,
-    infer: Optional[INFER_FUNC_TYPE] = None,
-    eval: Optional[EVAL_FUNC_TYPE] = None,
-    eval_configs: Optional[Union[TYPE_EVALUATION_CONFIG, List[TYPE_EVALUATION_CONFIG]]] = None,
+    results: Union[pd.DataFrame, List[Tuple[TYPE_EVALUATION_CONFIG, pd.DataFrame]]],
+    on: TEST_ON_TYPE = None,
```

Comment:
Based on current implementation logic, this should not be `None`, right?

Reply:
currently could be None, see [here](https://github.com/kolenaIO/kolena/pull/289/files#diff-b8784ed1fedc689303755b9c54e6936c49f2ab3715e1c140ac15a6dc402f9187R168-R169)